### PR TITLE
User button multi Function

### PR DESCRIPTION
**Pull request by gustavosinbandera1**
plz dont accept this PR now, I need to implement the needed functions to turn on/off device

### CODE REVIEW

File: components/HMI/src/HumanInterface.cpp [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+#include <cstdint>
+#include <cstring>
+#include <esp_log.h>
+
+namespace hmi {
+static xQueueHandle gpio_evt_queue = nullptr;
+// ----- Initialization and Default Values -----
+/**
+ * @brief Construct a new OneButton object but not (yet) initialize the IO pin.
+ */
```
Comment (jeandudey) [inline components/HMI/src/HumanInterface.cpp]: Move this to the header file.

File: components/HMI/src/HumanInterface.cpp [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    _gpio_btn = static_cast<gpio_num_t>(-1);
+}
+
+/**
+ * @brief Construct a new Button:: Button object
+ * 
+ * @param gpio 
+ * @param activeLow 
+ * @param pullupActive 
+ */
```
Comment (jeandudey) [inline components/HMI/src/HumanInterface.cpp]: Move this to the header file.

File: components/HMI/src/HumanInterface.cpp [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+Button::Button(gpio_num_t gpio, int activeLow, bool pullupActive)
+{
+} //Button
+
+// save function for click event
```
Comment (jeandudey) [inline components/HMI/src/HumanInterface.cpp]: This comment repeats what is on the doc comment.

File: components/HMI/src/HumanInterface.cpp [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+/**
+ * @brief save function for  click event
+ * 
+ * @param newFunction callback function
+ */
```
Comment (jeandudey) [inline components/HMI/src/HumanInterface.cpp]: Move to header file.

File: components/HMI/src/HumanInterface.cpp [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+namespace hmi {
+static xQueueHandle gpio_evt_queue = nullptr;
+// ----- Initialization and Default Values -----
+/**
+ * @brief Construct a new OneButton object but not (yet) initialize the IO pin.
+ */
+Button::Button()
+{
+    _gpio_btn = static_cast<gpio_num_t>(-1);
+}
```
Comment (jeandudey) [inline components/HMI/src/HumanInterface.cpp]: Why this constructor exists? It creates an invalid _gpio_btn. Does it has an use?

File: components/HMI/src/HumanInterface.cpp [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+
+/**
+ * @brief Construct a new Button:: Button object
+ * 
+ * @param gpio 
+ * @param activeLow 
+ * @param pullupActive 
+ */
+Button::Button(gpio_num_t gpio, int activeLow, bool pullupActive)
```
Comment (jeandudey) [inline components/HMI/src/HumanInterface.cpp]: `snakeCase` to `snake_case`.

File: components/HMI/src/HumanInterface.cpp [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+    _clickFunc = newFunction;
+} // attachClick
+
+/**
+ * @brief save function for parameterized click event
+ * 
+ * @param newFunction 
+ * @param parameter 
+ */
+void Button::attachClick(parameterizedCallbackFunction newFunction, void* parameter)
```
Comment (jeandudey) [inline components/HMI/src/HumanInterface.cpp]: `snakeCase` to `snake_case`.

File: components/HMI/src/HumanInterface.cpp [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+{
+    _paramClickFunc = newFunction;
+    _clickFuncParam = parameter;
+} // attachClick
+
+
+// save function for doubleClick event
+void Button::attachDoubleClick(callbackFunction newFunction)
```
Comment (jeandudey) [inline components/HMI/src/HumanInterface.cpp]: `snakeCase` to `snake_case`.

File: components/HMI/src/HumanInterface.cpp [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+ * @param parameter 
+ */
+void Button::attachClick(parameterizedCallbackFunction newFunction, void* parameter)
+{
+    _paramClickFunc = newFunction;
+    _clickFuncParam = parameter;
+} // attachClick
+
+
+// save function for doubleClick event
```
Comment (jeandudey) [inline components/HMI/src/HumanInterface.cpp]: Convert to documentation comment and move to header file.

File: components/HMI/src/HumanInterface.cpp [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+} // attachDoubleClick
+
+
+/**
+ * @brief save function for parameterized doubleClick event
+ * 
+ * @param newFunction 
+ * @param parameter 
+ */
+void Button::attachDoubleClick(parameterizedCallbackFunction newFunction, void* parameter)
```
Comment (jeandudey) [inline components/HMI/src/HumanInterface.cpp]: `snakeCase` to `snake_case`.

File: components/HMI/src/HumanInterface.cpp [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+    _doubleClickFunc = newFunction;
+} // attachDoubleClick
+
+
+/**
+ * @brief save function for parameterized doubleClick event
+ * 
+ * @param newFunction 
+ * @param parameter 
+ */
```
Comment (jeandudey) [inline components/HMI/src/HumanInterface.cpp]: Move to header file.

File: components/HMI/src/HumanInterface.cpp [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+void Button::attachDoubleClick(callbackFunction newFunction)
+{
+    _doubleClickFunc = newFunction;
+} // attachDoubleClick
+
+
+/**
+ * @brief save function for parameterized doubleClick event
+ * 
+ * @param newFunction
```
Comment (jeandudey) [inline components/HMI/src/HumanInterface.cpp]: Add parameter documentation.

File: components/HMI/src/HumanInterface.cpp [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+void Button::attachClick(callbackFunction newFunction)
+{
+    _clickFunc = newFunction;
+} // attachClick
+
+/**
+ * @brief save function for parameterized click event
+ * 
+ * @param newFunction 
+ * @param parameter
```
Comment (jeandudey) [inline components/HMI/src/HumanInterface.cpp]: Add parameters documentation.

File: components/HMI/src/HumanInterface.cpp [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+ */
+Button::Button(gpio_num_t gpio, int activeLow, bool pullupActive)
+{
+} //Button
+
+// save function for click event
+/**
+ * @brief save function for  click event
+ * 
+ * @param newFunction callback function
```
Comment (jeandudey) [inline components/HMI/src/HumanInterface.cpp]: Add parameters documentation.

File: components/HMI/src/HumanInterface.cpp [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+{
+    _gpio_btn = static_cast<gpio_num_t>(-1);
+}
+
+/**
+ * @brief Construct a new Button:: Button object
+ * 
+ * @param gpio 
+ * @param activeLow 
+ * @param pullupActive
```
Comment (jeandudey) [inline components/HMI/src/HumanInterface.cpp]: Add parameters documentation.

File: components/HMI/src/HumanInterface.h [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+    //virtual void run(void* arg);
+
+    // attach functions that will be called when button was pressed in the
+    // specified way.
+    void attachClick(callbackFunction newFunction);
+    void attachClick(parameterizedCallbackFunction newFunction, void* parameter);
+    void attachDoubleClick(callbackFunction newFunction);
+    void attachDoubleClick(parameterizedCallbackFunction newFunction, void* parameter);
+    void attachLongClick(callbackFunction newFunction);
+    void attachLongClick(parameterizedCallbackFunction newFunction, void* parameter);
```
Comment (jeandudey) [inline components/HMI/src/HumanInterface.h]: - `snakeCase` to `snake_case`.
- Move the documentation from the .cpp to here.

File: components/HMI/src/HumanInterface.h [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+    BTN_PRESS_1,
+    BTN_END_PROCESS,
+} state;
+
+class Button //: Task
+{
+public:
+    Button();
+    Button(gpio_num_t gpio, int active, bool pullupActive = true);
+    //virtual void run(void* arg);
```
Comment (jeandudey) [inline components/HMI/src/HumanInterface.h]: Unused code?

File: components/HMI/src/HumanInterface.h [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+    BTN_RELEASE_1,
+    BTN_PRESS_1,
+    BTN_END_PROCESS,
+} state;
+
+class Button //: Task
+{
+public:
+    Button();
+    Button(gpio_num_t gpio, int active, bool pullupActive = true);
```
Comment (jeandudey) [inline components/HMI/src/HumanInterface.h]: `snakeCase` to `snake_case`.

File: components/HMI/src/HumanInterface.h [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+
+static enum states {
+    BTN_RELEASE_0,
+    BTN_PRESS_0,
+    BTN_RELEASE_1,
+    BTN_PRESS_1,
+    BTN_END_PROCESS,
+} state;
+
+class Button //: Task
```
Comment (jeandudey) [inline components/HMI/src/HumanInterface.h]: - Class documentation
- Unused code? I mean `: Task`

File: components/HMI/src/HumanInterface.h [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+
+    /**
+   * @brief Call this function every time the input level has changed.
+   * Using this function no digital input pin is checked because the current
+   * level is given by the parameter.
+   */
+    void tick(bool level, unsigned long time_now);
+
+    bool isLongPressed();
+    int getPressedTicks();
```
Comment (jeandudey) [inline components/HMI/src/HumanInterface.h]: Shouldn't it return `TickType_t` (type safety)?

File: components/HMI/src/HumanInterface.h [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+
+    callbackFunction _doubleClickFunc = NULL;
+    parameterizedCallbackFunction _paramDoubleClickFunc = NULL;
+    void* _doubleClickFuncParam = NULL;
+
+
+    callbackFunction _longClickFunc = NULL;
+    parameterizedCallbackFunction _paramLongClickFunc = NULL;
+    void* _longClickFuncParam = NULL;
+
```
Comment (jeandudey) [inline components/HMI/src/HumanInterface.h]: - Use the m_ prefix for member variables (private).
- `snakeCase` to `snake_case`.
- Use doxygen comments.

File: components/HMI/src/UserButton.h [hunk 22]
```diff
@@ -0,0 +1,41 @@
+#ifndef HMI_USER_BUTTON_H
+#define HMI_USER_BUTTON_H
+
+#include "HumanInterface.h"
+#include "Task.h"
+#include "driver/gpio.h"
+
+namespace hmi {
+typedef void (*callbackFunction)(void);
+typedef void (*parameterizedCallbackFunction)(void*);
```
Comment (jeandudey) [inline components/HMI/src/UserButton.h]: New types should be SnakeCase to not confuse them with a function or variable.

File: components/HMI/src/HumanInterface.cpp (hunk 5 above)
Comment (gustavosinbandera1) [reply components/HMI/src/HumanInterface.cpp]: you right , I need to remove it

**Comment by luisan00** [comment]
This needs to be resolved guys 🇳🇮 🇻🇪 🇨🇴

**Pull request closed without merging**